examples/image.rs

```diff
@@ -56,82 +58,110 @@ fn run(dkr_ref: &reference::Reference, user: Option<String>, passwd: Option<Stri
+        }).and_then(|(dclient, manifest_kind)| {
+            let image = image.clone();
+            dclient
+                .get_manifest(&image, &version)
+                .and_then(move |manifest_body| {
+                    let layers = match manifest_kind {
+                        dkregistry::mediatypes::MediaTypes::ManifestV2S1Signed => {
+                            let m: dkregistry::v2::manifest::ManifestSchema1Signed =
+                        serde_json::from_slice(manifest_body.as_slice()).unwrap();
+                            m.get_layers()
+                        }
+                        dkregistry::mediatypes::MediaTypes::ManifestV2S2 => {
+                            let m: dkregistry::v2::manifest::ManifestSchema2 =
+                            serde_json::from_slice(manifest_body.as_slice()).unwrap();
+                            m.get_layers()
+                        }
+                        _ => return Err("unknown format".into()),
+                    };
+                    Ok((dclient, layers))
+                })
+        }).and_then(|(dclient, layers)| {
+            let image = image.clone();
+
+            println!("{} -> got {} layer(s)", &image, layers.len(),);
+
+            futures::stream::iter_ok::<_, dkregistry::errors::Error>(layers)
+                .and_then(move |layer| {
+                    let get_blob_future = dclient.get_blob(&image, &layer);
+                    get_blob_future.inspect(move |blob| {
+                        println!("Layer {}, got {} bytes.\n", layer, blob.len());
+                    })
+                }).collect()
+        });
+
+    let blobs = match tcore.run(futures) {
+        Ok(blobs) => blobs,
+        Err(e) => return Err(Box::new(e)),
+        _ => return Err("Couldn't get image".into()),
     };
 
-    println!(
-        "{} -> got {} layer(s), saving to directory {:?}",
-        image,
-        layers.len(),
-        version
-    );
-    std::fs::create_dir(&version)?;
-    let mut blobs: Vec<Vec<u8>> = vec![];
-
-    for (i, digest) in layers.iter().enumerate() {
-        let fut_presence = dclient.has_blob(&image, &digest)?;
-        let has_blob = tcore.run(fut_presence)?;
-        if !has_blob {
-            return Err(format!("missing layer {}", digest).into());
-        }
+    println!("Downloaded {} layers", blobs.len());
 
-        println!("Downloading layer {}...", digest);
-        let fut_out = dclient.get_blob(&image, &digest)?;
-        let out = tcore.run(fut_out)?;
-        println!(
-            "Layer {}/{}, got {} bytes.\n",
-            i + 1,
-            layers.len(),
-            out.len()
-        );
-        blobs.push(out);
+    // FIXME: use async io
+    let path = &format!("{}:{}", &image, &version).replace("/", "_");
+    let path = std::path::Path::new(&path);
+    if path.exists() {
+        return Err(format!("path {:?} already exists, exiting", &path).into());
     }
+    std::fs::create_dir(&path).unwrap();
```

Comment:
For reference: https://docs.rs/tokio/0.1.12/tokio/fs/fn.create_dir_all.html

We are still using the old `tokio_core` everywhere in this repo, so I think we can leave this FIXME in place and tackle that once we plainly switch to the new `tokio` crate.

Reply:
I added a TODO here